File: java_client_src/src/ihs/apcs/spacebattle/networking/MwnpMessage.java
package ihs.apcs.spacebattle.networking;

import com.google.gson.Gson;

import ihs.apcs.spacebattle.*;
import ihs.apcs.spacebattle.commands.ShipCommand;
import ihs.apcs.spacebattle.util.StringMap;
import ihs.apcs.spacebattle.util.StringStringMap;

public class MwnpMessage {
	private int[] ids;
	private String command;
	private Object data;
	
	private static StringMap<Class<?>> cmdDataTypes;
	
	static {
		cmdDataTypes = new StringMap<Class<?>>();
		cmdDataTypes.put("\"MWNL2_ASSIGNMENT\"", double.class);
		cmdDataTypes.put("\"REQUEST\"", StringStringMap.class);
		cmdDataTypes.put("\"ENV\"", Environment.class);
		cmdDataTypes.put("\"ERROR\"", ErrorData.class);
	}
	
	public MwnpMessage(int[] ids, String command, Object data) {
		this.ids = ids;
		this.command = command;
		this.data = data;
	}
	
	public MwnpMessage(int[] ids, ShipCommand cmd) throws IllegalArgumentException, IllegalAccessException {
		this.ids = ids;
		this.command = "SCMD";
		this.data = cmd.getMessage();
	}
	
	public MwnpMessage(int[] ids, RegistrationData regData) {
		this.ids = ids;
		this.command = "REGISTER";
		this.data = regData;
	}
	
	public int getSenderId() { return ids[0]; }
	public int getReceiverId() { return ids[1]; }
	public String getCommand() { return command; }
	public Object getData() { return data; }
	
	/**
	 * Parses JSON text into a MwnpMessage.
	 * This process is non-trivial because the network messages contain mixed types.
	 * It is assumed that messages look like this:
	 *   ##[sender, receiver],COMMANDNAME,DATA
	 * where ## is the length of the message in bytes.  The message must then be parsed
	 *   as three separate JSON objects. 
	 * @param messageText the string representation of the message
	 * @return the @MwnpMessage object created from this text
	 */
	public static MwnpMessage parseMessage(String messageText) {
		String id = messageText.substring(0, messageText.indexOf(']') + 1);
		String command = messageText.substring(id.length() + 1, messageText.indexOf('\"', messageText.indexOf('\"') + 1) + 1);

		Gson gson = new Gson();
		Class<?> dataType = cmdDataTypes.get(command);

		if (dataType != null) {
			String data = messageText.substring(id.length() + command.length() + 2);
			
			return new MwnpMessage (
					gson.fromJson(id, int[].class),
					gson.fromJson(command, String.class), 
					gson.fromJson(data, dataType));
		} else {
			return new MwnpMessage (
					gson.fromJson(id, int[].class),
					gson.fromJson(command, String.class), 
					null);			
		}
	}
	
	/**
	 * Converts this message into its JSON representation.
	 * @return the JSON representation of this message
	 */
	public String toJsonString() {
		Gson gson = new Gson();
		StringBuilder build = new StringBuilder();
		
		build.append(gson.toJson(this.ids));
		build.append(",");
		build.append(gson.toJson(this.command));
		if (this.data != null) {
			build.append(",");
			build.append(gson.toJson(this.data));
		}
		
		int msgSize = build.length();
		build.insert(0, msgSize);
		
		return build.toString();
	}
}

File: cs_client_src/networking/MwnpMessage.cs
using System;
using System.Linq;
using System.Text;
using Newtonsoft.Json;

public class MwnpMessage {
	private int[] ids;
	private string command;
	private object data;
	
	private static StringMap<Type> cmdDataTypes = new StringMap<Type>() {
		{ "\"MWNL2_ASSIGNMENT\"", typeof(double) },
		{ "\"REQUEST\"", typeof(StringStringMap) },
		{ "\"ERROR\"", typeof(ErrorData) }
	};
	
	public static void RegisterGameType(string gameName)
	{
		switch (gameName)
		{
			case "BaubleHunt":
				cmdDataTypes.Add("\"ENV\"", typeof(Environment<BaubleHuntGameInfo>));
				break;
			case "KingOfTheBubble":
				cmdDataTypes.Add("\"ENV\"", typeof(Environment<KingOfTheBubbleGameInfo>));
				break;
			case "DiscoveryQuest":
				cmdDataTypes.Add("\"ENV\"", typeof(Environment<DiscoveryQuestGameInfo>));
				break;
			case "TheHungerBaubles":
				cmdDataTypes.Add("\"ENV\"", typeof(Environment<TheHungerBaublesGameInfo>));
				break;
			default:
				cmdDataTypes.Add("\"ENV\"", typeof(Environment<BasicGameInfo>));
				break;
		}
	}
	
	public MwnpMessage(int[] ids, string command, object data) {
		this.ids = ids;
		this.command = command;
		this.data = data;
	}
	
	public MwnpMessage(int[] ids, ShipCommand cmd) {
		this.ids = ids;
		this.command = "SCMD";
		this.data = cmd.getMessage();
	}
	
	public MwnpMessage(int[] ids, RegistrationData regData) {
		this.ids = ids;
		this.command = "REGISTER";
		this.data = regData;
	}
	
	public int getSenderId() { return ids[0]; }
	public int getReceiverId() { return ids[1]; }
	public string getCommand() { return command; }
	public object getData() { return data; }
	
	public static MwnpMessage parseMessage(string messageText) {
		string id = messageText.Substring(0, messageText.IndexOf(']') + 1);
		string command = messageText.Substring(id.Length + 1, messageText.IndexOf('\"', messageText.IndexOf('\"') + 1) - id.Length);

		Type dataType = cmdDataTypes.ContainsKey(command) ? cmdDataTypes[command] : null;

		if (dataType != null) {
			string data = messageText.Substring(id.Length + command.Length + 2);		
			
			return new MwnpMessage(
					JsonConvert.DeserializeObject<int[]>(id),
					JsonConvert.DeserializeObject<string>(command), 
					JsonConvert.DeserializeObject(data, dataType));
		} else {
			return new MwnpMessage (
					JsonConvert.DeserializeObject<int?[]>(id).Select(i => i > 0 ? i.Value : -1).ToArray(), 
					JsonConvert.DeserializeObject<string>(command), 
					null);			
		}
	}
	
	/**
	 * Converts this message into its JSON representation.
	 * @return the JSON representation of this message
	 */
	public string toJsonstring() {
		StringBuilder build = new StringBuilder();
		
		build.Append(JsonConvert.SerializeObject(this.ids));
		build.Append(",");
		build.Append(JsonConvert.SerializeObject(this.command));
		if (this.data != null) {
			build.Append(",");
			build.Append(JsonConvert.SerializeObject(this.data));
		}
		
		int msgSize = build.Length;
		build.Insert(0, msgSize);
		
		return build.ToString();
	}

	override public string ToString()
	{
		return toJsonstring();
	}
}
